modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java

```diff
@@ -161,15 +159,17 @@ public PeekCursor<IndexRow> scan(@Nullable BinaryTuplePrefix lowerBound, @Nullab
 
             SortedIndexRowKey upper = createBound(upperBound, includeUpper);
 
-            return new ScanCursor<IndexRow>(lower, sortedIndexTree) {
+            return new ScanCursor<IndexRow>(lower) {
+                private final BinaryTupleComparator comparator = indexTree.getBinaryTupleComparator();
```

Comment:
Could it happen that the constructor captures one tree instance, but the comparator will be taken from another one due to a race? How about taking the tree instance before creating the scan cursor explicitly and then using it? It seems that it will require to restore the explicit argument in the constructor to pass a tree.